h2/src/main/org/h2/util/DateTimeUtils.java

```diff
@@ -1043,9 +1062,12 @@ public static long dateValueFromDenormalizedDate(long year, long month, int day)
      * @return the date value
      */
     public static long dateValueFromDate(long ms) {
-        Calendar cal = getCalendar();
-        cal.setTimeInMillis(ms);
-        return dateValueFromCalendar(cal);
+        ms += getTimeZone().getOffset(ms);
+        long absoluteDay = ms / 86_400_000;
+        if (ms < 0 && (absoluteDay * 86_400_000 != ms)) {
```

Comment:
what is this logic for? and pull out the constant and give it a name. thanks.

Reply:
This logic is required for dates before 1970. Division in Java rounds result toward to zero, but here we need result rounded toward negative infinity.

I put comments here and in the another method.